core/libp2p/connection/connection.hpp

```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_CONNECTION_HPP
+#define KAGOME_CONNECTION_HPP
+
+#include <vector>
+
+#include <boost/optional.hpp>
+#include "common/result.hpp"
+#include "libp2p/basic/writable.hpp"
+#include "libp2p/common/network_message.hpp"
+#include "libp2p/common/peer_info.hpp"
+#include "libp2p/multi/multiaddress.hpp"
+
+namespace libp2p::connection {
+  /**
+   * Point-to-point link to the other peer
+   */
+  class Connection : public basic::Writable {
+    /**
+     * Get addresses, which are observed with an underlying transport
+     * @return collection of such addresses
+     */
+    virtual std::vector<multi::Multiaddress> getObservedAdrresses() const = 0;
+
+    /**
+     * Get information about the peer this connection connects to
+     * @return peer information if set, none otherwise
+     */
+    virtual boost::optional<common::PeerInfo> getPeerInfo() const = 0;
+
+    /**
+     * Set information about the peer this connection connects to
+     * @param info to be set
+     */
+    virtual void setPeerInfo(const common::PeerInfo &info) = 0;
```

Comment:
Not really sure it's a good interface. Probably PeerInfo might be set via ctor only, otherwise the abstraction becomes very fragile

Reply:
We agreed to implement an interface, which is 99% consistent with the specification, and this is how it's done there. If during implementation we find that your statement is true, it won't be hard to change it